File: frontend/src/utils/trains.ts
/**
 * Group an ordered list of events into "trains" (clusters of closely
 * spaced events). Pure function — no I/O, no state. The same algorithm
 * exists in backend/analysis/trains.py; the two MUST stay in sync. See
 * tests/trains_parity.* for the fixture used to verify that.
 *
 * The algorithm is a single forward pass: walk consecutive pairs, start
 * a candidate train when the gap is small enough, append while it stays
 * small, close it when the gap widens. After the pass, drop trains
 * shorter than the minimum count or duration, then optionally merge
 * adjacent trains separated by less than min_inter_train_ms.
 *
 * Designed so point events (APs, synaptic peak times) and extended
 * events (epileptiform bursts) share one code path: a point event has
 * tStart === tEnd === t.
 */

export type TrainMetric = 'gap' | 'peak_to_peak'

export interface TrainEvent {
  /** Reference time, seconds. For point events this is the only time
   *  that matters; for extended events it is typically the peak. */
  t: number
  /** Event onset, seconds. Defaults to t for point events. */
  tStart?: number
  /** Event offset, seconds. Defaults to t for point events. */
  tEnd?: number
}

export interface TrainParams {
  metric: TrainMetric
  /** Maximum inter-event interval (ms) for two events to belong to the
   *  same train. Interpretation depends on metric. */
  maxIeiMs: number
  /** Minimum number of events to count as a train. */
  minCount: number
  /** Minimum total train duration (ms). 0 disables the floor. */
  minDurationMs?: number
  /** Minimum silence (ms) between two trains; trains separated by less
   *  than this get merged after initial grouping. 0 disables. */
  minInterTrainMs?: number
}

export interface TrainSummary {
  id: number
  startS: number
  endS: number
  durationMs: number
  nEvents: number
  meanIeiMs: number | null
  intraFreqHz: number | null
  memberIndices: number[]
}

export interface TrainResult {
  /** Same length as the input event list. null = isolated event. */
  trainIdByIndex: (number | null)[]
  trainSummaries: TrainSummary[]
}

const EMPTY: TrainResult = { trainIdByIndex: [], trainSummaries: [] }

function ieiMs(
  prev: TrainEvent,
  next: TrainEvent,
  metric: TrainMetric,
): number {
  if (metric === 'peak_to_peak') {
    return (next.t - prev.t) * 1000
  }
  // gap (end of prev → start of next)
  const prevEnd = prev.tEnd ?? prev.t
  const nextStart = next.tStart ?? next.t
  return (nextStart - prevEnd) * 1000
}

function summarize(
  events: TrainEvent[],
  members: number[],
  id: number,
  metric: TrainMetric,
): TrainSummary {
  const first = events[members[0]]
  const last = events[members[members.length - 1]]
  const startS = first.tStart ?? first.t
  const endS = last.tEnd ?? last.t
  const durationMs = (endS - startS) * 1000
  let meanIeiMs: number | null = null
  let intraFreqHz: number | null = null
  if (members.length >= 2) {
    let sum = 0
    for (let i = 1; i < members.length; i++) {
      sum += ieiMs(events[members[i - 1]], events[members[i]], metric)
    }
    meanIeiMs = sum / (members.length - 1)
    if (meanIeiMs > 0) intraFreqHz = 1000 / meanIeiMs
  }
  return {
    id,
    startS,
    endS,
    durationMs,
    nEvents: members.length,
    meanIeiMs,
    intraFreqHz,
    memberIndices: members.slice(),
  }
}

export function computeTrains(
  events: TrainEvent[] | null | undefined,
  params: TrainParams,
): TrainResult {
  if (!events || events.length === 0) return EMPTY
  const n = events.length
  if (n < 2) return { trainIdByIndex: new Array(n).fill(null), trainSummaries: [] }

  const minCount = Math.max(2, Math.floor(params.minCount))
  const minDurationMs = Math.max(0, params.minDurationMs ?? 0)
  const minInterTrainMs = Math.max(0, params.minInterTrainMs ?? 0)
  const maxIeiMs = params.maxIeiMs
  const metric = params.metric

  // Forward pass: candidate clusters.
  const candidates: number[][] = []
  let current: number[] = [0]
  for (let i = 1; i < n; i++) {
    const dt = ieiMs(events[i - 1], events[i], metric)
    if (dt <= maxIeiMs) {
      current.push(i)
    } else {
      candidates.push(current)
      current = [i]
    }
  }
  candidates.push(current)

  // Filter by min count and min duration.
  let kept = candidates.filter((m) => {
    if (m.length < minCount) return false
    if (minDurationMs > 0) {
      const f = events[m[0]]
      const l = events[m[m.length - 1]]
      const dur = ((l.tEnd ?? l.t) - (f.tStart ?? f.t)) * 1000
      if (dur < minDurationMs) return false
    }
    return true
  })

  // Optional merge: trains separated by < min_inter_train_ms.
  if (minInterTrainMs > 0 && kept.length > 1) {
    const merged: number[][] = [kept[0]]
    for (let i = 1; i < kept.length; i++) {
      const prev = merged[merged.length - 1]
      const a = events[prev[prev.length - 1]]
      const b = events[kept[i][0]]
      const gapMs = ieiMs(a, b, metric)
      if (gapMs < minInterTrainMs) {
        merged[merged.length - 1] = prev.concat(kept[i])
      } else {
        merged.push(kept[i])
      }
    }
    kept = merged
  }

  const trainIdByIndex: (number | null)[] = new Array(n).fill(null)
  const trainSummaries: TrainSummary[] = []
  kept.forEach((members, idx) => {
    for (const i of members) trainIdByIndex[i] = idx
    trainSummaries.push(summarize(events, members, idx, metric))
  })

  return { trainIdByIndex, trainSummaries }
}

File: backend/analysis/trains.py
"""Group an ordered list of events into "trains" (clusters of closely
spaced events). Pure function — no FastAPI, no numpy required.

Twin of ``frontend/src/utils/trains.ts``. The two implementations MUST
produce identical output for the same input; ``tests/trains_parity.py``
verifies this against a fixture JSON. If you change the algorithm
here, change the TS version too (or vice versa) and re-run the test.

Designed so point events (APs, synaptic peak times) and extended events
(epileptiform bursts) share one code path: a point event has
``t_start == t_end == t``.
"""

from __future__ import annotations

from typing import Optional


def _iei_ms(prev: dict, nxt: dict, metric: str) -> float:
    if metric == "peak_to_peak":
        return (float(nxt["t"]) - float(prev["t"])) * 1000.0
    # gap
    prev_end = float(prev.get("t_end", prev["t"]))
    next_start = float(nxt.get("t_start", nxt["t"]))
    return (next_start - prev_end) * 1000.0


def _summarize(
    events: list[dict],
    members: list[int],
    train_id: int,
    metric: str,
) -> dict:
    first = events[members[0]]
    last = events[members[-1]]
    start_s = float(first.get("t_start", first["t"]))
    end_s = float(last.get("t_end", last["t"]))
    duration_ms = (end_s - start_s) * 1000.0
    mean_iei_ms: Optional[float] = None
    intra_freq_hz: Optional[float] = None
    if len(members) >= 2:
        total = 0.0
        for i in range(1, len(members)):
            total += _iei_ms(events[members[i - 1]], events[members[i]], metric)
        mean_iei_ms = total / (len(members) - 1)
        if mean_iei_ms > 0:
            intra_freq_hz = 1000.0 / mean_iei_ms
    return {
        "id": train_id,
        "start_s": start_s,
        "end_s": end_s,
        "duration_ms": duration_ms,
        "n_events": len(members),
        "mean_iei_ms": mean_iei_ms,
        "intra_freq_hz": intra_freq_hz,
        "member_indices": list(members),
    }


def group_into_trains(
    events: list[dict],
    *,
    metric: str = "gap",
    max_iei_ms: float,
    min_count: int = 2,
    min_duration_ms: float = 0.0,
    min_inter_train_ms: float = 0.0,
) -> tuple[list[Optional[int]], list[dict]]:
    """Cluster consecutive events whose IEI is below ``max_iei_ms``.

    ``events`` must be sorted by ``t``. Each event is a dict with at
    least ``t`` (seconds); extended events also carry ``t_start`` and
    ``t_end``. Returns ``(train_id_per_event, train_summaries)`` where
    ``train_id_per_event[i]`` is None for isolated events.
    """
    if not events:
        return [], []
    n = len(events)
    if n < 2:
        return [None] * n, []

    min_count = max(2, int(min_count))
    min_duration_ms = max(0.0, float(min_duration_ms))
    min_inter_train_ms = max(0.0, float(min_inter_train_ms))

    # Forward pass.
    candidates: list[list[int]] = []
    current: list[int] = [0]
    for i in range(1, n):
        dt = _iei_ms(events[i - 1], events[i], metric)
        if dt <= max_iei_ms:
            current.append(i)
        else:
            candidates.append(current)
            current = [i]
    candidates.append(current)

    # Filter by min count and min duration.
    kept: list[list[int]] = []
    for m in candidates:
        if len(m) < min_count:
            continue
        if min_duration_ms > 0:
            f = events[m[0]]
            l = events[m[-1]]
            dur = (float(l.get("t_end", l["t"])) - float(f.get("t_start", f["t"]))) * 1000.0
            if dur < min_duration_ms:
                continue
        kept.append(m)

    # Optional merge of close-but-separated trains.
    if min_inter_train_ms > 0 and len(kept) > 1:
        merged: list[list[int]] = [kept[0]]
        for i in range(1, len(kept)):
            prev = merged[-1]
            a = events[prev[-1]]
            b = events[kept[i][0]]
            gap_ms = _iei_ms(a, b, metric)
            if gap_ms < min_inter_train_ms:
                merged[-1] = prev + kept[i]
            else:
                merged.append(kept[i])
        kept = merged

    train_id_per_event: list[Optional[int]] = [None] * n
    summaries: list[dict] = []
    for idx, members in enumerate(kept):
        for i in members:
            train_id_per_event[i] = idx
        summaries.append(_summarize(events, members, idx, metric))

    return train_id_per_event, summaries
